# Report showing alignment details and stats

```{r  echo=FALSE, warning=FALSE, message=FALSE}
	library(reshape2)
	library(ggplot2)

	metric_table <- data[['metric_table']]

	print(names(metric_table))

	make_stacked_barchart_from_df <- function(df, id, other_vars) {
		reduced_df <- df[,c(id, other_vars)]
		melt_df <- melt(reduced_df, id.vars = id )
		g <- ggplot(melt_df, aes_string(x=id, y="value", fill="variable"))
		g + geom_bar(stat='identity', position=position_dodge()) + theme(axis.text.x = element_text(angle = 45, hjust = 1))
	}

	make_scatterplot_from_df <- function(df, var1, var2) {
		g <- ggplot(df, aes_string(x=var1, y=var2))
		g + geom_point()
	}

	cat("Reads before and after trimming:")
	make_stacked_barchart_from_df(
		df = metric_table, 
        id = "sample", 
        other_vars = c("initial_total_sequences", "final_total_sequences")
    )



	cat("Reads mapped to genome, those uniquely mapping, and how many overlap with features:")
    make_stacked_barchart_from_df(
		df = metric_table, 
        id = "sample", 
        other_vars = c("number_reads", "Uniquely_mapped_reads_number", "aligned_to_feature")
    )

    cat("Do samples with few aligned genes have low quality scores?:")
    make_scatterplot_from_df(
		df = metric_table,
		var1 = "aligned_to_feature",
		var2 = "initial_weigthed_qual_per_sequence"
	)

    cat("Number of replicate reads")
    make_stacked_barchart_from_df(
		df = metric_table, 
        id = "sample", 
        other_vars = c("initial_total_sequences", "number_duplicatedreads.estimated.")
    )

```